#!/usr/bin/env python

import sys


_SEX = {
    '0': 'UNKNOWN',
    '1': 'MALE',
    '2': 'FEMALE'
}

_KAR_SEX = {
    '0': None,
    '1': 'XY',
    '2': 'XX'
}

FNAME_TEMPLATE = 'ALL.chr{}.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz'

def create_individuals(ind_info, yml_fhand):
    text = []
    text.append('individuals:')
    for ind in ind_info:
        text.append('{}- id: {}'.format(' '*2, ind['Individual ID']))
        text.append('{}name: {}'.format(' '*4, ind['Individual ID']))
        if ind['Paternal ID'] != '0':
            text.append('{}father:'.format(' '*4))
            text.append('{}- id: {}'.format(' '*6, ind['Paternal ID']))
        if ind['Maternal ID'] != '0':
            text.append('{}mother:'.format(' '*4))
            text.append('{}- id: {}'.format(' '*6, ind['Maternal ID']))
        text.append('{}sex: {}'.format(' '*4, _SEX[ind['Gender']]))
        text.append('{}karyotypicSex: {}'.format(' '*4, _KAR_SEX[ind['Gender']]))
        text.append('{}population:'.format(' '*4, ind['Population']))
        text.append('{}- name: {}'.format(' '*6, ind['Population']))
        text.append('{}samples:'.format(' '*4))
        text.append('{}- id: {}'.format(' '*6, ind['Individual ID']))
    return '\n'.join(text)

def create_samples(ind_info, yml_fhand):
    text = []
    text.append('samples:')
    for ind in ind_info:
        text.append('{}- id: {}'.format(' '*2, ind['Individual ID']))
        text.append('{}individualId: {}'.format(' '*4, ind['Individual ID']))
    return '\n'.join(text)

def create_files(yml_fhand):
    text = []
    text.append('files:')
    for chrom in range(1, 23) + ['X', 'Y', 'MT']:
        text.append('{}- id: {}'.format(' '*2, FNAME_TEMPLATE.format(chrom)))
        text.append('{}path: {}'.format(' '*4, '/data'))
    return '\n'.join(text)

def main():
    ped_fpath = sys.argv[1]
    yml_fpath = sys.argv[2]

    ped_fhand = open(ped_fpath, 'r')
    yml_fhand = open(yml_fpath, 'w')

    header = ped_fhand.readline().strip().split('\t')
    ind_info = [{k: v for k, v in zip(header, line.strip().split('\t'))} for line in ped_fhand]

    yml_fhand.write('---\n')
    yml_fhand.write('# WARNING: AUTOGENERATED CONTENT\n')
    yml_fhand.write('id: 1000g\n')
    yml_fhand.write('name: 1000 Genomes phase 3\n')
    yml_fhand.write('description: The 1000 Genomes Project\n')
    yml_fhand.write(create_individuals(ind_info, yml_fhand) + '\n')
    yml_fhand.write(create_files(yml_fhand) + '\n')

    ped_fhand.close()
    yml_fhand.close()


if __name__ == '__main__':
    sys.exit(main())