
################################################################################
#' AnalysisAlignmentClient methods
#' @include commons.R

#' @description This function implements the OpenCGA calls for managing AnalysisAlignment
#' @param OpencgaR an object OpencgaR generated using initOpencgaR and/or opencgaLogin
#' @seealso \url{https://github.com/opencb/opencga/wiki} and the RESTful API documentation
#' \url{http://bioinfo.hpc.cam.ac.uk/opencga/webservices/}
#' @export


setMethod("analysisalignmentClient", "OpencgaR", function(OpencgaR, action, params=NULL, ...) {
    category <- "analysis/alignment"
    switch(action,
        # Endpoint: /{apiVersion}/analysis/alignment/bwa/run
        # @param study: study.
        # @param jobId: Job ID. It must be a unique string within the study. An id will be autogenerated automatically if not provided.
        # @param jobDependsOn: Comma separated list of existing job ids the job will depend on.
        # @param jobDescription: Job description.
        # @param jobTags: Job tags.
        # @param data: BWA parameters.
        runBwa=fetchOpenCGA(object=OpencgaR, category=category, categoryId=NULL, subcategory="bwa", subcategoryId=NULL,
                action="run", params=params, httpMethod="POST", as.queryParam=NULL, ...),
        # Endpoint: /{apiVersion}/analysis/alignment/coverage/index/run
        # @param study: study.
        # @param jobId: Job ID. It must be a unique string within the study. An id will be autogenerated automatically if not provided.
        # @param jobDependsOn: Comma separated list of existing job ids the job will depend on.
        # @param jobDescription: Job description.
        # @param jobTags: Job tags.
        # @param data: Coverage computation parameters.
        runCoverageIndex=fetchOpenCGA(object=OpencgaR, category=category, categoryId=NULL,
                subcategory="coverage/index", subcategoryId=NULL, action="run", params=params, httpMethod="POST",
                as.queryParam=NULL, ...),
        # Endpoint: /{apiVersion}/analysis/alignment/coverage/query
        # @param file: File ID.
        # @param study: Study [[user@]project:]study where study and project can be either the ID or UUID.
        # @param region: Comma separated list of regions 'chr:start-end, e.g.: 2,3:63500-65000.
        # @param gene: Comma separated list of genes, e.g.: BCRA2,TP53.
        # @param offset: Offset to extend the region, gene or exon at up and downstream.
        # @param onlyExons: Only exons are taking into account when genes are specified.
        # @param range: Range of coverage values to be reported. Minimum and maximum values are separated by '-', e.g.: 20-40 (for coverage values greater or equal to 20 and less or equal to 40). A single value means to report coverage values less or equal to that value.
        # @param windowSize: Window size for the region coverage (if a coverage range is provided, window size must be 1).
        # @param splitResults: Split results into regions (or gene/exon regions).
        queryCoverage=fetchOpenCGA(object=OpencgaR, category=category, categoryId=NULL, subcategory="coverage",
                subcategoryId=NULL, action="query", params=params, httpMethod="GET", as.queryParam=c("file"), ...),
        # Endpoint: /{apiVersion}/analysis/alignment/coverage/ratio
        # @param file1: Input file #1 (e.g. somatic file).
        # @param file2: Input file #2 (e.g. germline file).
        # @param study: Study [[user@]project:]study where study and project can be either the ID or UUID.
        # @param skipLog2: Do not apply Log2 to normalise the coverage ratio.
        # @param region: Comma separated list of regions 'chr:start-end, e.g.: 2,3:63500-65000.
        # @param gene: Comma separated list of genes, e.g.: BCRA2,TP53.
        # @param offset: Offset to extend the region, gene or exon at up and downstream.
        # @param onlyExons: Only exons are taking into account when genes are specified.
        # @param windowSize: Window size for the region coverage (if a coverage range is provided, window size must be 1).
        # @param splitResults: Split results into regions (or gene/exon regions).
        ratioCoverage=fetchOpenCGA(object=OpencgaR, category=category, categoryId=NULL, subcategory="coverage",
                subcategoryId=NULL, action="ratio", params=params, httpMethod="GET", as.queryParam=c("file1","file2"),
                ...),
        # Endpoint: /{apiVersion}/analysis/alignment/coverage/stats
        # @param file: File ID.
        # @param gene: Comma separated list of genes, e.g.: BCRA2,TP53.
        # @param study: Study [[user@]project:]study where study and project can be either the ID or UUID.
        # @param threshold: Only regions whose coverage depth is under this threshold will be reported.
        statsCoverage=fetchOpenCGA(object=OpencgaR, category=category, categoryId=NULL, subcategory="coverage",
                subcategoryId=NULL, action="stats", params=params, httpMethod="GET", as.queryParam=c("file","gene"),
                ...),
        # Endpoint: /{apiVersion}/analysis/alignment/deeptools/run
        # @param study: study.
        # @param jobId: Job ID. It must be a unique string within the study. An id will be autogenerated automatically if not provided.
        # @param jobDependsOn: Comma separated list of existing job ids the job will depend on.
        # @param jobDescription: Job description.
        # @param jobTags: Job tags.
        # @param data: Deeptools parameters.
        runDeeptools=fetchOpenCGA(object=OpencgaR, category=category, categoryId=NULL, subcategory="deeptools",
                subcategoryId=NULL, action="run", params=params, httpMethod="POST", as.queryParam=NULL, ...),
        # Endpoint: /{apiVersion}/analysis/alignment/fastqc/run
        # @param study: study.
        # @param jobId: Job ID. It must be a unique string within the study. An id will be autogenerated automatically if not provided.
        # @param jobDependsOn: Comma separated list of existing job ids the job will depend on.
        # @param jobDescription: Job description.
        # @param jobTags: Job tags.
        # @param data: FastQC parameters.
        runFastqc=fetchOpenCGA(object=OpencgaR, category=category, categoryId=NULL, subcategory="fastqc",
                subcategoryId=NULL, action="run", params=params, httpMethod="POST", as.queryParam=NULL, ...),
        # Endpoint: /{apiVersion}/analysis/alignment/index/run
        # @param study: study.
        # @param jobId: Job ID. It must be a unique string within the study. An id will be autogenerated automatically if not provided.
        # @param jobDependsOn: Comma separated list of existing job ids the job will depend on.
        # @param jobDescription: Job description.
        # @param jobTags: Job tags.
        # @param data: Alignment index params.
        runIndex=fetchOpenCGA(object=OpencgaR, category=category, categoryId=NULL, subcategory="index",
                subcategoryId=NULL, action="run", params=params, httpMethod="POST", as.queryParam=NULL, ...),
        # Endpoint: /{apiVersion}/analysis/alignment/query
        # @param limit: Number of results to be returned.
        # @param skip: Number of results to skip.
        # @param count: Get the total number of results matching the query. Deactivated by default.
        # @param file: File ID.
        # @param study: Study [[user@]project:]study where study and project can be either the ID or UUID.
        # @param region: Comma separated list of regions 'chr:start-end, e.g.: 2,3:63500-65000.
        # @param gene: Comma separated list of genes, e.g.: BCRA2,TP53.
        # @param offset: Offset to extend the region, gene or exon at up and downstream.
        # @param onlyExons: Only exons are taking into account when genes are specified.
        # @param minMappingQuality: Minimum mapping quality.
        # @param maxNumMismatches: Maximum number of mismatches.
        # @param maxNumHits: Maximum number of hits.
        # @param properlyPaired: Return only properly paired alignments.
        # @param maxInsertSize: Maximum insert size.
        # @param skipUnmapped: Skip unmapped alignments.
        # @param skipDuplicated: Skip duplicated alignments.
        # @param regionContained: Return alignments contained within boundaries of region.
        # @param forceMDField: Force SAM MD optional field to be set with the alignments.
        # @param binQualities: Compress the nucleotide qualities by using 8 quality levels.
        # @param splitResults: Split results into regions (or gene/exon regions).
        query=fetchOpenCGA(object=OpencgaR, category=category, categoryId=NULL, subcategory=NULL, subcategoryId=NULL,
                action="query", params=params, httpMethod="GET", as.queryParam=c("file"), ...),
        # Endpoint: /{apiVersion}/analysis/alignment/samtools/run
        # @param study: study.
        # @param jobId: Job ID. It must be a unique string within the study. An id will be autogenerated automatically if not provided.
        # @param jobDependsOn: Comma separated list of existing job ids the job will depend on.
        # @param jobDescription: Job description.
        # @param jobTags: Job tags.
        # @param data: Samtoolstools parameters.
        runSamtools=fetchOpenCGA(object=OpencgaR, category=category, categoryId=NULL, subcategory="samtools",
                subcategoryId=NULL, action="run", params=params, httpMethod="POST", as.queryParam=NULL, ...),
        # Endpoint: /{apiVersion}/analysis/alignment/stats/info
        # @param file: File ID.
        # @param study: Study [[user@]project:]study where study and project can be either the ID or UUID.
        infoStats=fetchOpenCGA(object=OpencgaR, category=category, categoryId=NULL, subcategory="stats",
                subcategoryId=NULL, action="info", params=params, httpMethod="GET", as.queryParam=c("file"), ...),
        # Endpoint: /{apiVersion}/analysis/alignment/stats/run
        # @param study: study.
        # @param jobId: Job ID. It must be a unique string within the study. An id will be autogenerated automatically if not provided.
        # @param jobDependsOn: Comma separated list of existing job ids the job will depend on.
        # @param jobDescription: Job description.
        # @param jobTags: Job tags.
        # @param data: Alignment stats params.
        runStats=fetchOpenCGA(object=OpencgaR, category=category, categoryId=NULL, subcategory="stats",
                subcategoryId=NULL, action="run", params=params, httpMethod="POST", as.queryParam=NULL, ...),
    )
})